src/HexagonAlignLoads.cpp

```diff
@@ -0,0 +1,217 @@
+        // We reason only in terms of lanes. Each lane is a vector element.
+        // We want to know the following.
+        //    1. if the base of buffer + ramp->base (i.e. the index) are aligned.
+        //    2. if not, then how many lanes off an aligned address are they (returned in base_mod).
+        //    3. if 2, then we create two loads and slice_vector them.
+        //    4. rem_mod is used if the ramp base is 64*x + 65 and lanes is 64, then we are not
+        //       modulus_remainder.remainder lanes off, but we are only 1 lane off.
+        int lanes = ramp->lanes;
+        Expr base = ramp->base;
+        // We know the base itself puts us off an aligned address by base_lanes_off
+        // number of lanes.
+        int base_lanes_off = mod_imp(host_alignment, lanes);
+        ModulusRemainder mod_rem = get_alignment_info(base);
+        if (mod_rem.modulus == 1 && mod_rem.remainder == 0) {
+            // We can't reason about alignment.
+            return AlignCheck::NoResult;
+        } else {
+            int base_mod = base_lanes_off + mod_imp(mod_rem.modulus, lanes);
+            int rem_mod = mod_imp(mod_rem.remainder, lanes);
+            if (!(base_mod + rem_mod)) {
+                return AlignCheck::Aligned;
+            } else {
+                if (lanes_off) *lanes_off = base_mod + rem_mod;
+                return AlignCheck::Unaligned;
+            }
+        }
+    }
+    void visit(const Load *op) {
+        debug(4) << "HexagonAlignLoads: Working on " << (Expr) op << "..\n";
+        Expr index = mutate(op->index);
+        if (op->type.is_vector()) {
+            bool external = op->image.defined();
+            if (external) {
+                debug(4) << "HexagonAlignLoads: Not dealing with an external image\n";
+                debug(4) << (Expr) op << "\n";
+                expr = op;
+                return;
+            } else {
+                const Ramp *ramp = index.as<Ramp>();
+                const IntImm *stride = ramp ? ramp->stride.as<IntImm>() : NULL;
+                // We will work only on natural vectors supported by the target.
+                if (ramp->lanes != natural_vector_lanes(op->type)) {
+                    expr = op;
+                    return;
+                }
+                if (ramp && stride && stride->value == 1) {
+                    int lanes = ramp->lanes;
+                    // If this is a parameter, the base_alignment should be host_alignment.
+                    // This cannot be an external image because we have already checked for
+                    // it. Otherwise, this is an internal buffers that is always aligned
+                    // to the natural vector width.
+                    int base_alignment =
+                        op->param.defined() ? op->param.host_alignment() : vector_size;
+                    int lanes_off;
+                    AlignCheck ac = get_alignment_check(ramp, base_alignment, &lanes_off);
```

Comment:
If you went with my suggestion above, this would become:

    bool known_alignment = get_alignment_check(ramp->base, base_alignment, &lanes_off);
    if (known_alignment && lanes_off != 0) {
       ....
    } else {
      expr = op;
    }